matrix/plugins/glitch/selectors.py

```diff
@@ -0,0 +1,108 @@
+import functools
+import logging
+import random
+from typing import List, Any
+
+import enforce
+from juju.model import Model
+from juju.application import Application
+from juju.unit import Unit
+
+from matrix.utils import Singleton
+
+
+_marker = object()
+log = logging.getLogger("glitch")
+
+
+class _Selectors(dict, metaclass=Singleton):
+    def decorate(self, f):
+        wrapper = enforce.runtime_validation(f)
```

Comment:
I don't understand the sentiment that adding type checking implies we should give up Python entirely.  As you mentioned, we were already starting to do ad-hoc type checking, and there are plenty of other reasons to use Python over Scala.

Type hints are now a standard feature of Python, albeit optional, and it we should use it when it makes sense.  One of Python's strengths has always been the willingness to adopt ideas from other languages and paradigms when they make sense, while still allowing flexibility to not use them when they don't.